python/ray/data/dataset.py

```diff
@@ -2868,75 +2832,72 @@ def to_tf(
+
+        def generator():
+            for batch in self.iter_tf_batches(
                 prefetch_blocks=prefetch_blocks,
                 batch_size=batch_size,
-                batch_format="pandas",
                 drop_last=drop_last,
                 local_shuffle_buffer_size=local_shuffle_buffer_size,
                 local_shuffle_seed=local_shuffle_seed,
             ):
-                if label_column:
-                    targets = convert_pandas_to_tf_tensor(batch[[label_column]])
-                    if targets.ndim == 2 and targets.shape[1] == 1:
-                        targets = tf.squeeze(targets, axis=1)
-                    batch.pop(label_column)
+                assert isinstance(batch, dict)
+                features = get_columns_from_batch(batch, columns=feature_columns)
+                labels = get_columns_from_batch(batch, columns=label_columns)
+                yield features, labels
 
-                features = None
-                if feature_columns is None:
-                    features = convert_pandas_to_tf_tensor(batch)
-                elif isinstance(feature_columns, list):
-                    if all(isinstance(column, str) for column in feature_columns):
-                        features = convert_pandas_to_tf_tensor(batch[feature_columns])
-                    elif all(isinstance(columns, list) for columns in feature_columns):
-                        features = tuple(
-                            convert_pandas_to_tf_tensor(batch[columns])
-                            for columns in feature_columns
-                        )
-                    else:
-                        raise ValueError(
-                            "Expected `feature_columns` to be a list of strings or a "
-                            "list of lists."
-                        )
-                elif isinstance(feature_columns, dict):
-                    features = {
-                        key: convert_pandas_to_tf_tensor(batch[columns])
-                        for key, columns in feature_columns.items()
-                    }
-                else:
-                    raise ValueError(
-                        "Expected `feature_columns` to be a list or a dictionary, "
-                        f"but got a `{type(feature_columns).__name__}` instead."
-                    )
-
-                if label_column:
-                    yield features, targets
-                else:
-                    yield features
+        feature_type_spec = get_type_spec(self.schema(), columns=feature_columns)
+        label_type_spec = get_type_spec(self.schema(), columns=label_columns)
+        output_signature = (feature_type_spec, label_type_spec)
 
         dataset = tf.data.Dataset.from_generator(
-            make_generator, output_signature=output_signature
+            generator, output_signature=output_signature
         )
 
-        return dataset
+        return prepare_dataset_shard(dataset)
```

Comment:
Do we need to do this here? This is pretty specific to Train and may need to be overridden by the user (e.g. if they want to disable or modify TF prefetching). Personally would prefer to just have the user to call this in a separate line.

Reply:
I think we should keep it. IIRC distributed training doesn't work without it.

@clarkzinzow what do you think?

Reply:
We need to disable the `tf.data` autoshard policy at a minimum, otherwise the iterator will be broken, so that's something that we should always do. I'm also not convinced that enabling the `tf.data`-level prefetching is a good idea for the reasons that I gave [here](https://github.com/ray-project/ray/pull/28861#discussion_r984063217).

I would push for one of the following:
1. Don't call `prepare_dataset_shard`, and instead disable the autoshard policy on the TensorFlow Dataset returned here:
```suggestion
        options = tf.data.Options()
        options.experimental_distribute.auto_shard_policy = (
            tf.data.experimental.AutoShardPolicy.OFF
        )
        return dataset.with_options(options)
```
2. Modify `prepare_dataset_shard` to only disable the autoshard policy (i.e. no `tf.data` prefetching), and always invoke it here.

I'd vote for (2), but we can go with (1) if we think that changing the prefetching semantics of `prepare_dataset_shard` is risky.

Reply:
I think it's safer to go with (1) and re-evaluate (2) as a follow-up.